provider/openstack/cinder.go

```diff
@@ -0,0 +1,443 @@
+		// size parameter is in GB. It is actually GiB.
+		Size: int(math.Ceil(float64(size / 1024))),
+		Name: tag.String(),
+	})
+	if err != nil {
+		return storage.Volume{}, err
+	}
+	return cinderToJujuVolume(tag, &resp.Volume), err
+}
+
+func (ga *gooseAdapter) DeleteVolume(volId string) error {
+	return ga.cinder.DeleteVolume(volId)
+}
+
+func (ga *gooseAdapter) GetVolumesSimple() ([]storage.Volume, error) {
+
+	resp, err := ga.cinder.GetVolumesSimple()
+	if err != nil {
+		return nil, err
+	}
+
+	vols := make([]storage.Volume, len(resp.Volumes))
+	for idx, vol := range resp.Volumes {
+		vols[idx] = cinderToJujuVolume(names.VolumeTag{}, &vol)
+	}
+
+	return vols, nil
+}
+
+func (ga *gooseAdapter) AttachVolume(serverId, volumeId, mountPoint string) (storage.VolumeAttachment, error) {
+	attachment, err := ga.nova.AttachVolume(serverId, volumeId, mountPoint)
+	if err != nil {
+		return storage.VolumeAttachment{}, err
+	}
+	return novaToJujuVolumeAttachment(attachment), nil
+}
+
+func (ga *gooseAdapter) VolumeStatusNotifier(volId, status string, numAttempts int, waitDur time.Duration) <-chan error {
+	return ga.cinder.VolumeStatusNotifier(volId, status, numAttempts, waitDur)
+}
+
+func (ga *gooseAdapter) ListVolumeAttachments(serverId string) ([]storage.VolumeAttachment, error) {
+	novaVolAttachments, err := ga.nova.ListVolumeAttachments(serverId)
+	if err != nil {
+		return nil, err
+	}
+
+	volumes := make([]storage.VolumeAttachment, len(novaVolAttachments))
+	for volIdx, volAttach := range novaVolAttachments {
+		volumes[volIdx] = novaToJujuVolumeAttachment(&volAttach)
+	}
+
+	return volumes, nil
+}
+
+func (ga *gooseAdapter) DetachVolume(serverId, attachmentId string) error {
+	return ga.nova.DetachVolume(serverId, attachmentId)
+}
+
+const (
+	openstackCfgKeyUrl        = "auth-url"
+	openstackCfgKeyUser       = "username"
+	openstackCfgKeyPassword   = "password"
+	openstackCfgKeyAccessKey  = "access-key"
+	openstackCfgKeySecretKey  = "secret-key"
+	openstackCfgKeyRegion     = "region"
+	openstackCfgKeyTenantName = "tenant-name"
+)
+
+// func NewOpenstackStorageConfig(url, user, secrets, region, tenantName string) (*storage.Config, error) {
+
+// 	return storage.NewConfig(
+// 		"cinder",
+// 		CinderProviderType,
+// 		map[string]interface{}{
+// 			openstackCfgKeyUrl:        url,
+// 			openstackCfgKeyUser:       user,
+// 			openstackCfgKeySecrets:    secrets,
+// 			openstackCfgKeyRegion:     region,
+// 			openstackCfgKeyTenantName: tenantName,
+// 		},
+// 	)
+// }
+
+func cinderToJujuVolume(tag names.VolumeTag, volume *cinder.Volume) storage.Volume {
+	var size uint64
+	size = uint64(int64(volume.Size * 1024))
+	return storage.Volume{
+		VolumeId: volume.ID,
+		Size:     size,
+		Tag:      tag,
+	}
+}
+
+func novaToJujuVolumeAttachment(volAttach *nova.VolumeAttachment) storage.VolumeAttachment {
+	return storage.VolumeAttachment{
+		Volume:     names.NewVolumeTag(volAttach.VolumeId),
+		Machine:    names.NewMachineTag(volAttach.ServerId),
```

Comment:
likewise.

tags correspond to Juju's internal names for entities ("0", "1", etc.), not whatever OpenStack or AWS says.